File: goens/pkg/enslib/conn.go
package enslib

import (
	"encoding/binary"
	"net"
)

func SendENSMsg(conn net.Conn, msg *ENSMsg) (err error) {
	sendData, err := msg.Encode()
	if err != nil {
		return
	}

	_, err = conn.Write(sendData)
	if err != nil {
		return
	}

	return nil
}

func RecvENSMsg(conn net.Conn) (msg *ENSMsg, err error) {
	dataSize := binary.Size(ENSMsgData{})

	recvData := make([]byte, dataSize)

	_, err = conn.Read(recvData)
	if err != nil {
		return
	}

	msg, err = DecodeENSMsg(recvData)
	if err != nil {
		return
	}

	return msg, nil
}

File: rs-ens/src/enslib/conn.rs
use std::{
    io::{Read, Write},
    net::TcpStream,
};

use super::msg::{ENSMsg, ENS_MGS_SIZE};
use std::io::Result;

pub fn send_ens_msg(stream: &mut TcpStream, msg: &ENSMsg) -> Result<()> {
    let data = msg.encode();

    stream.write_all(data.as_slice())?;

    return Ok(());
}

pub fn recv_ens_msg(stream: &mut TcpStream) -> Option<ENSMsg> {
    let mut recv_data: [u8; 101] = [0u8; ENS_MGS_SIZE];

    if let Err(_) = stream.read_exact(&mut recv_data) {
        return None;
    }

    return ENSMsg::decode(&recv_data.to_vec());
}
